File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Model/Expression/VariableExpression.java
package Model.Expression;

import Util.Dictionary.MyIDictionary;
import Util.Heap.IHeap;

public class VariableExpression implements IExpression {
    private String name;

    public VariableExpression(String name){
        this.name = name;
    }

    @Override
    public int evaluate(MyIDictionary<String,Integer> symbolTable, IHeap heap) throws ExpressionException {
        if(symbolTable.containsKey(this.name))
            return symbolTable.get(this.name);

        throw new ExpressionException("No such symbol in the symbol table!");
    }

    @Override
    public String toString() {
        return "(" + this.name + ")";
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Model/Expression/VariableExpression.cs
using ToyLanguage.Util.Dictionary;

namespace ToyLanguage.Model.Expression
{
    class VariableExpression : IExpression
    {
        private readonly string id;

        public VariableExpression(string id)
        {
            this.id = id;
        }

        public int Evaluate(MyIDictionary<string, int> symbolTable)
        {
            if (!symbolTable.TryGetValue(id, out int value)) throw new ExpressionException("Unassigned variable used!");

            return value;
        }

        public override string ToString()
        {
            return "VariableExpression(" + id + ")";
        }
    }
}
